File: java/src/main/java/com/epi/PalindromeNumberTest.java
package com.epi;

// @pg_import:4
import com.epi.utils.AbstractTestOptions;
import com.epi.utils.AbstractTestStream;
import com.epi.utils.JsonTestOptions;
import com.epi.utils.TestType;

// @pg_ignore:1
import static com.epi.PalindromeNumber.isPalindromeNumber;

public class PalindromeNumberTest {
  // @pg_ignore
  // @pg_include:PalindromeNumber.java
  // @pg_end
  private static void unitTest(AbstractTestOptions options, String description,
                               int input, boolean expected) {
    AbstractTestStream stream = options.getStream();
    stream.startTest(TestType.NORMAL, description);
    stream.registerInput(input);
    stream.registerExpectedOutput(expected);

    try {
      boolean result = isPalindromeNumber(input);
      stream.registerUserOutput(result, result == expected);
    }
    catch (NullPointerException e) {
      stream.registerNullPointerException();
    }
    catch (Exception e) {
      stream.registerUnhandledException();
    }

    stream.endTest();
  }

  public static void directedTests(AbstractTestOptions options) {
    options.startTests(0, "Palindrome number");

    unitTest(options, "Negative value test #1", -1, false);
    unitTest(options, "Negative value test #2", -1001, false);
    unitTest(options, "Negative value test #3", Integer.MIN_VALUE + 1, false);
    unitTest(options, "Negative value test #4", Integer.MIN_VALUE, false);
    unitTest(options, "Zero value test", 0, true);
    unitTest(options, "Single number test #1", 1, true);
    unitTest(options, "Single number test #2", 9, true);
    unitTest(options, "Odd-length palindrome test", 52125, true);
    unitTest(options, "Even-length palindrome test", 521125, true);
    unitTest(options, "Odd-length non-palindrome test", 52105, false);
    unitTest(options, "Even-length non-palindrome test", 520125, false);
    unitTest(options, "Big number test #1", Integer.MAX_VALUE - 1, false);
    unitTest(options, "Big number test #2", Integer.MAX_VALUE, false);

    options.endTests();
  }

  public static void main(String[] args) {
    directedTests(new JsonTestOptions(System.out));
  }

}

File: cpptest/palindrome-numberTest.cc
// Copyright (c) 2016 Elements of Programming Interviews. All rights reserved.
#include <limits>

#include "test_toolkit/test-options.h"
// @pg_ignore:1
#include "palindrome-number.cc"
// @pg_include:palindrome-number.cc

using std::numeric_limits;

void UnitTest(TestSentry::Ptr& sentry, const char *description, int x, bool expected) {
  TestStream::Ptr stream = sentry->GetTestStream(TestType::NORMAL, description);
  stream->RegisterInput(x);
  stream->RegisterExpectedOutput(expected);
  try {
    bool result = IsPalindromeNumber(x);
    stream->RegisterUserOutput(result, result == expected);
  }
  catch (...) {
    stream->RegisterUnhandledException();
  }
}

void DirectedTests(const TestOptions& options) {
  TestSentry::Ptr sentry = options.GetTestSentry(3, "Palindrome numbers");

  UnitTest(sentry, "Negative value test #1", -1, false);
  UnitTest(sentry, "Negative value test #2", -1001, false);
  UnitTest(sentry, "Negative value test #3", numeric_limits<int>::min() + 1, false);
  UnitTest(sentry, "Negative value test #4", numeric_limits<int>::min(), false);
  UnitTest(sentry, "Zero value test", 0, true);
  UnitTest(sentry, "Single number test #1", 1, true);
  UnitTest(sentry, "Single number test #2", 9, true);
  UnitTest(sentry, "Odd-length palindrome test", 52125, true);
  UnitTest(sentry, "Even-length palindrome test", 521125, true);
  UnitTest(sentry, "Odd-length non-palindrome test", 52105, false);
  UnitTest(sentry, "Even-length non-palindrome test", 520125, false);
  UnitTest(sentry, "Big number test #1", numeric_limits<int>::max() - 1, false);
  UnitTest(sentry, "Big number test #2", numeric_limits<int>::max(), false);
}

int main(int argc, char **argv) {
  DirectedTests(TestOptions(&cout));
  return 0;
}
